docs/how-to/hip_porting_guide.rst

```diff
@@ -0,0 +1,715 @@
+even when ``hip_runtime_api.h`` or ``hip_runtime.h`` are included.
+
+The HIP include paths and platform macros (``__HIP_PLATFORM_AMD__`` or
+``__HIP_PLATFORM_NVIDIA__``) must be passed to the compiler.
+
+``hipconfig`` can help in finding the necessary options, for example on an AMD
+platform:
+
+.. code-block:: bash
+
+  hipconfig --cpp_config
+   -D__HIP_PLATFORM_AMD__= -I/opt/rocm/include
+
+``nvcc`` includes some headers by default. ``hipcc`` does not include
+default headers, and instead all required files must be explicitly included.
+
+The ``hipify`` tool automatically converts ``cuda_runtime.h`` to
+``hip_runtime.h``, and it converts ``cuda_runtime_api.h`` to
+``hip_runtime_api.h``, but it may miss nested headers or macros.
+
+warpSize
+================================================================================
+
+Code should not assume a warp size of 32 or 64, as that is not portable between
+platforms and architectures. The ``warpSize`` built-in should be used in device
+code, while the host can query it during runtime via the device properties. See
+the :ref:`HIP language extension for warpSize <warp_size>` for information on
+how to write portable wave-aware code.
+
+Lane masks bit-shift
+================================================================================
+
+A thread in a warp is also called a lane, and a lane mask is a bitmask where
+each bit corresponds to a thread in a warp. A bit is 1 if the thread is active,
+0 if it's inactive. Bit-shift operations are typically used to create lane masks
+and on AMD GPUs the ``warpSize`` can differ between different architectures,
+that's why it's essential to use correct bitmask type, when porting code.
+
+Example:
+
+.. code-block:: cpp
+
+  // Get the thread's position in the warp
+  unsigned int laneId = threadIdx.x % warpSize;
+
+  // Use lane ID for bit-shift
+  val & ((2 << (threadIdx.x % warpSize) )-1 );
+
+  // Shift 32 bit integer with val variable
+  WarpReduce::sum( (val < warpSize) ? (1 << val) : 0);
+
+Lane masks are 32-bit integer types as this is the integer precision that C 
+assigns to such constants by default. GCN/CDNA architectures have a warp size of
+64, :code:`threadIdx.x % warpSize` and :code:`val` in the example may obtain 
+values greater than 31. Consequently, shifting by such values would  clear the 
+32-bit register to which the shift operation is applied to. For AMD
+architectures, a straightforward fix could look as follows:
+
+.. code-block:: cpp
+  
+  // Get the thread's position in the warp
+  unsigned int laneId = threadIdx.x % warpSize;
+
+  // Use lane ID for bit-shift
+  val & ((2ull << (threadIdx.x % warpSize) )-1 );
+
+  // Shift 64 bit integer with val variable
+  WarpReduce::sum( (val < warpSize) ? (1ull << val) : 0);
+
+For portability reasons, one should rather introduce appropriately
+typed placeholders as shown below:
+
+.. code-block:: cpp
+
+  #if defined(__GFX8__) || defined(__GFX9__)
```

Comment:
Why wouldn't this be based on the warpsize specified rather than the GPU?

Reply:
Because the warpSize parameter will be non compile time variable and here we need compile time variable.
More details under the issue: https://github.com/ROCm/ROCm/issues/4121

Reply:
Would it be worth adding how to obtain a compile time variable?